core/src/main/scala/scalaz/Generator.scala

```diff
@@ -1,29 +1,32 @@
 package scalaz
 
 /**
- * A `Generator[C]` is a container of elements, and which knows how to efficiently apply a [[scalaz.Reducer]]
+ * A `Generator[C, Elem]` is a container of elements, and which knows how to efficiently apply a [[scalaz.Reducer]]
  * to extract an answer by combining elements. A `Reducer` may supply efficient left-to-right and
  * right-to-left reduction strategies that a `Generator` may avail itself of.
  */
-trait Generator[C[_]] {
-  def reduce[E, M](r: Reducer[E, M], c: C[E]): M = to(r, r.zero, c)
+trait Generator[C, Elem] {
 
-  def to[E, M](r: Reducer[E, M], m: M, c: C[E]): M = r.append(m, reduce(r, c))
+  def reduce[E, M](e: Elem => E, c: C)(implicit r: Reducer[E, M], m: Monoid[M]): M = to(e, m.zero, c)
 
-  def from[E, M](r: Reducer[E, M], c: C[E], m: M): M = r.append(reduce(r, c), m)
+  def to[E, M](e: Elem => E, m: M, c: C)(implicit r: Reducer[E, M], M: Monoid[M]): M = r.append(m, reduce(e, c))
+
+  def from[E, M](e: Elem => E, c: C, m: M)(implicit r: Reducer[E, M], M: Monoid[M]): M = r.append(reduce(e, c), m)
```

Comment:
I know these are not introduced in this PR but what's the purpose of `to` and `from`? Why are they two distinct methods? The signature is the same, modulo parameter order, and the default implementation suspicious (again pre-existent, not introduced in this PR). Also, since `m: M` is taken as an argument, it seems strange to also require `Monoid[M]`.

Reply:
Good point question. I'm guessing this is to allows optimal implementation for some instance where `append` is costly. The only instance I found that actually overwrite this method is for `Free`: https://github.com/ekmett/ekmett.github.com/blob/master/haskell/monoids/Data/Generator/Free.hs#L109

The unnecessary `Monoid` constraint on `to` is so that it can be implemented in term of `reduce` for the case where the later method is used as minimal definition.

We could drop the constraint and have `to` as the only minimal definition. (but this would also likely means we also drop the `from` method altogether, unless we keep the constraint but that would be weird...)

@ekmett, do you have any recommendation on the matter? (is `Generator.mapFrom` useful in practice?)

Reply:
Regarding

> which knows how to efficiently apply a [[scalaz.Reducer]]
> to extract an answer by combining elements. A `Reducer` may supply efficient left-to-right and
> right-to-left reduction strategies that a `Generator` may avail itself of.

the `Generator` only knows how to efficiently traverse a container. The `Reducer` may have an efficient reduction in one direction, but not the other. The `Generator` would choose the optimal strategy for the container, but that might have pathological performance for the reducer. It might therefore be better to have one-directional reducers.

Reply:
The main purpose of the Generator/Reducer classes is to be a lightweight general purpose map-reduce framework. So the Generator must have the ability to reduce in parallel, aggregating from both sides. Caller is responsible to chose Generator/reducer pair that work well with each other.
(edit: moved comment)

Reply:
If the caller is responsible for matching of reducers to generators, then shouldn't both be explicit (as opposed to implicit)?